libc/unistd_file.c

```diff
@@ -0,0 +1,774 @@
+	TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, read(fd, buf, sizeof(LINE1) - 1));
+	TEST_ASSERT_EQUAL_STRING(LINE1, buf);
+	/* assert EOF */
+	TEST_ASSERT_EQUAL_INT(0, read(fd, buf, sizeof(LINE1) - 1));
+	TEST_ASSERT_EMPTY(buf);
+
+	/* fallback relative to file end */
+	TEST_ASSERT_EQUAL_INT(0, (off_t)lseek(fd, -(sizeof(LINE1) - 1), SEEK_END));
+	/* read line */
+	TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, read(fd, buf, sizeof(LINE1) - 1));
+	TEST_ASSERT_EQUAL_STRING(LINE1, buf);
+	/* assert EOF */
+	TEST_ASSERT_EQUAL_INT(0, read(fd, buf, sizeof(LINE1) - 1));
+	TEST_ASSERT_EMPTY(buf);
+}
+
+/* test lseek() outside of the file */
+TEST(unistd_file, file_lseek_pastfile)
+{
+	/* offsets for SEEK_SET, SEEK_CUR and SEEK_END tests of past-the-file lseek() */
+	off_t setoff = 2 * (sizeof(LINE1) - 1);
+	off_t curoff = setoff + sizeof(LINE1) - 1;
+	off_t endoff = curoff + sizeof(LINE1) - 1;
+
+	/* fill file with content */
+	TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+
+	/* reopen file */
+	fd2 = open(filename, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fd2);
+	{
+		/* set fd2 past current EOF position using SEEK_SET */
+		TEST_ASSERT_EQUAL_INT(setoff, lseek(fd2, setoff, SEEK_SET));
+		/* fill empty gap between fd and fd2 and add new data */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, write(fd, LINE2, sizeof(LINE2) - 1));
+		/* assert correct read from fd2 */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, read(fd2, buf, sizeof(LINE2) - 1));
+		TEST_ASSERT_EQUAL_STRING(LINE2, buf);
+
+		/* set fd2 past current EOF position using SEEK_CUR */
+		TEST_ASSERT_GREATER_OR_EQUAL(curoff, lseek(fd2, sizeof(LINE1) - 1, SEEK_CUR));
+		/* fill empty gap between fd and fd2 and add new data */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, write(fd, LINE2, sizeof(LINE2) - 1));
+		/* assert correct read from fd2 */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, read(fd2, buf, sizeof(LINE2) - 1));
+		TEST_ASSERT_EQUAL_STRING(LINE2, buf);
+
+		/* set fd2 past current EOF position using SEEK_END */
+		TEST_ASSERT_GREATER_OR_EQUAL(endoff, lseek(fd2, sizeof(LINE1) - 1, SEEK_END));
+		/* fill empty gap between fd and fd2 and add new data */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, write(fd, LINE2, sizeof(LINE2) - 1));
+		/* assert correct read from fd2 */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, read(fd2, buf, sizeof(LINE2) - 1));
+		TEST_ASSERT_EQUAL_STRING(LINE2, buf);
+	}
+	TEST_ASSERT_EQUAL_INT(0, close(fd2));
+}
+
+/* test lseek() with offsets that result in overall negative offset in file */
+TEST(unistd_file, file_lseek_negative)
+{
+	TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+
+	/* <posix incmpliance> lseek() error with negative offset + strange errno
```

Comment:
Isn't this fixed already?

Reply:
Yes. Key bug was removed. Wrong errno is still returned (E2BIG instead of EINVAL)

I will try to pinpoint the cause. Thanks!